File: leetcode/medium/0875-koko-eating-bananas/solution.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1;
        int high = *max_element(piles.begin(), piles.end());
        int ans = high;

        while (low <= high){
            int k = low + (high-low)/2;
            long long hours = 0;

            for (int pile : piles){
                hours += (pile+k-1)/k;
            }

            if (hours <= h){
                ans = k;
                high = k - 1;
            }

            else {
                low = k+1;
            }
        }

        return ans;
    }
};

File: leetcode/medium/0875-koko-eating-bananas/solution.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int n = piles.length;
        int left = 1;
        int right = 0;
        for (int pile : piles){
            right = Math.max(right, pile);
        }

        while (left < right){
            int mid = left + (right-left)/2;
            int hours = 0;
            
            for (int pile : piles) hours += (mid+pile-1)/mid;

            if (hours <= h) right = mid;
            else left = mid+1;
        }
        return left;
    }
}